google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java

```diff
@@ -195,6 +198,160 @@ public void tearDown() {
     mockSpanner.removeAllExecutionTimes();
   }
 
+  @Test
+  public void testPoolMaintainer_whenInactiveTransactions_removeSessionsFromPool() {
+    InactiveTransactionRemovalOptions inactiveTransactionRemovalOptions =
+        InactiveTransactionRemovalOptions.newBuilder()
+            .setExecutionTimeThreshold(
+                Duration.ofMillis(5L)) // anything more than 5s will be long-running
+            .setRecurrenceDuration(Duration.ofSeconds(15L))
+            .build();
+    SessionPoolOptions sessionPoolOptions =
+        SessionPoolOptions.newBuilder()
+            .setMinSessions(1)
+            .setMaxSessions(1) // to ensure there is 1 session and pool is 100% utilized
+            .setCloseIfInactiveTransactions()
+            .setInactiveTransactionRemovalOptions(inactiveTransactionRemovalOptions)
+            .build();
+    spanner =
+        SpannerOptions.newBuilder()
+            .setProjectId(TEST_PROJECT)
+            .setDatabaseRole(TEST_DATABASE_ROLE)
+            .setChannelProvider(channelProvider)
+            .setCredentials(NoCredentials.getInstance())
+            .setSessionPoolOption(sessionPoolOptions)
+            .build()
+            .getService();
+    DatabaseClientImpl client =
+        (DatabaseClientImpl)
+            spanner.getDatabaseClient(DatabaseId.of(TEST_PROJECT, TEST_INSTANCE, TEST_DATABASE));
+    Instant initialExecutionTime = client.pool.poolMaintainer.lastExecutionTime;
+
+    try (TransactionManager manager = client.transactionManager()) {
+      TransactionContext transaction = manager.begin();
+      while (true) {
+        try {
+          // Simulate a delay of 20s to ensure that the below transaction is a long-running one.
+          // We require to wait for 20s so that the main thread executes at-least once every 10s
+          // As per this test, anything which takes more than 5s is long-running
+          mockSpanner.setExecuteSqlExecutionTime(
+              SimulatedExecutionTime.ofMinimumAndRandomTime(
+                  (int) Duration.ofSeconds(20).toMillis(), 0));
+          transaction.executeUpdate(UPDATE_STATEMENT);
+          manager.commit();
```

Comment:
This should fail, as the session has been returned to the pool. We should make sure we take measures that once we have forced the session back into the pool and someone is still using it, they get a reasonable error message telling them what went wrong. This will randomly succeed or fail in real life, depending on whether the session has been garbage collected by the backend or not.

I also wonder how hard it will fail if the session actually has been garbage collected, because the client also contains a retry-loop for `SessionNotFoundException`. So in that case the following will happen:
1. The session has been removed from the pool and garbage collected by the backend.
2. Trying to use it will lead to a `SessionNotFoundException`.
3. The client library will try to replace the session that was used by the transaction with a new one. But the session that was used has already been removed. What happens in that case?

Reply:
> This should fail, as the session has been returned to the pool.

This particular test did not fail since the session is not garbage collected in the backend. I can add another test where I throw a `SessionNotFoundException` and assert the behaviour. This particular test will continue to test the scenario where the session gets cleaned up on the front end. Is there any other behaviour that should be asserted for this case when the front-end returns the session while the backend hasn't garbage collected?

> We should make sure we take measures that once we have forced the session back into the pool and someone is still using it, they get a reasonable error message telling them what went wrong.

I will verify this through a new test, but I guess bubbling up a general `ResourceNotFound` / `SessionNotFound` error would be good enough? We originally thought of adding some messaging like `Transaction has been closed as it was running for more than 60 minutes. For long running transactions use batch or partitioned transactions` . But I think this is one of the reasons why a session is closed and we shouldn't be bubbling this to clients in all cases.

> The client library will try to replace the session that was used by the transaction with a new one. But the session that was used has already been removed. What happens in that case?

In the `replaceSession` flow we invalidate the previously used session. There is a check to do nothing in case the session is already in the closed state, hence I don't think we need anymore change here. I'll check through a test whether the flow here is expected or not. Overall, for the customer the transaction will still pass since we will replace with a new session.

Reply:
I think this is good. My initial mistake was thinking that the session was forced back to the pool, but we are not doing that. We are removing it from the pool alltogether.

Reply:
We need to think this through a bit more. Currently, if the `commit()` succeeds without any errors, the following happens:
1. The session is removed from the pool by the cleanup task.
2. The transaction continues, as it is not really inactive. It was just executing an RPC that was slow.
3. The `commit` succeeds and the `TransactionManager` is closed. This releases the session back into the pool. This again leaves the session pool in the following inconsistent state:
4.1. `SessionPool#allSessions` does not contain the session anymore.
4.2. `SessionPool#sessions` (the list of sessions currently checked in to the pool) does contain the session.

So we need to get this consistent, by either:
1. Making sure that a session that has been cleaned up by the background task cannot in any way be returned to the pool, even if the transaction that had it succeeded.
2. OR: Make sure that a session that had been cleaned up but that turns out to be alive after all can be fully returned to the pool (that is; the 'remove' action is completely undone)

Reply:
We concluded to favour the approach #1 where an exception never surfaces to the customer in case the session is recycled on client side. This is also consistent with what's done today when session is garbage collected on backend and we have a retry loop (after replacing client side session). 

I have added the necessary handling in the `Session#close()` method to ensure a session that is removed from the pool is not added back to the pool during close. Similar handling is already present in case of `SessionNotFound` exception.